source/d3d8to9_device.cpp

```diff
@@ -2134,3 +2152,21 @@ void Direct3DDevice8::ApplyClipPlanes()
 		index++;
 	}
 }
+
+template <typename T> void ReleaseShaderSet(std::unordered_set<DWORD>& ShaderHandles) {
+	for (auto Handle : ShaderHandles)
+	{
+		if (Handle == 0)
+			continue;
+		reinterpret_cast<T>(Handle)->Release();
+	}
+	ShaderHandles.clear();
+}
+
+void Direct3DDevice8::ReleaseShaders()
+{
+	ReleaseShaderSet<IDirect3DPixelShader9*>(PixelShaderHandles);
+	SetPixelShader(0);
+	ReleaseShaderSet<IDirect3DVertexShader9*>(VertexShaderHandles);
```

Comment:
Doesn't this leak the vertex declaration objects now? I think `VertexShaderHandles` needs to track the actual handles generated in `CreateVertexShader` and then this can simply call `DestroyVertexShader` on them (same thing for pixel shader handles). That will also do the `SetVertexShader(0)` calls internally already and thus avoid some code duplication.

Reply:
It does, thank you for noticing that! The host application I was using was manually deleting vertex shaders/declarations but not pixel shaders so I never noticed it in testing. I made the change you recommended.